lib/posix-socket/socket.c

```diff
@@ -0,0 +1,639 @@
+
+	file = posix_socket_file_get(sock);
+	if (unlikely(PTRISERR(file))) {
+		ret = PTR2ERR(file);
+		goto EXIT_ERR;
+	}
+
+	/* Send a structured message to a socket */
+	ret = posix_socket_sendmsg(file, msg, flags);
+
+	vfscore_put_file(file->vfs_file);
+
+	if (unlikely((ret < 0) && (ret != -EAGAIN)))
+		goto EXIT_ERR;
+
+	trace_posix_socket_sendmsg_ret(ret);
+	return ret;
+EXIT_ERR:
+	PSOCKET_ERR("sendmsg on socket %d failed: %d\n", sock, ret);
+	trace_posix_socket_sendmsg_err(ret);
+	return ret;
+}
+
+UK_TRACEPOINT(trace_posix_socket_sendto, "%d %p %d %d %p %d",
+	      int, const void *, size_t, int,
+	      const struct sockaddr *, socklen_t);
+UK_TRACEPOINT(trace_posix_socket_sendto_ret, "%d", int);
+UK_TRACEPOINT(trace_posix_socket_sendto_err, "%d", int);
+
+UK_SYSCALL_R_DEFINE(ssize_t, sendto, int, sock, const void *, buf, size_t, len,
+		    int, flags, const struct sockaddr *, dest_addr,
+		    socklen_t, addrlen)
+{
+	struct posix_socket_file *file;
+	int ret;
+
+	trace_posix_socket_sendto(sock, buf, len, flags, dest_addr, addrlen);
+
+	file = posix_socket_file_get(sock);
+	if (unlikely(PTRISERR(file))) {
+		ret = PTR2ERR(file);
+		goto EXIT_ERR;
+	}
+
+	/* Send to an address over a socket */
+	ret = posix_socket_sendto(file, buf, len, flags,
+		dest_addr, addrlen);
+
+	vfscore_put_file(file->vfs_file);
+
+	if (unlikely((ret < 0) && (ret != -EAGAIN)))
+		goto EXIT_ERR;
+
+	trace_posix_socket_sendto_ret(ret);
+	return ret;
+EXIT_ERR:
+	PSOCKET_ERR("sendto on socket %d failed: %d\n", sock, ret);
+	trace_posix_socket_sendto_err(ret);
+	return ret;
+}
+
+ssize_t send(int sock, const void *buf, size_t len, int flags)
+{
+	return sendto(sock, buf, len, flags, NULL, 0);
+}
+
+UK_TRACEPOINT(trace_posix_socket_socketpair, "%d %d %d %p", int, int, int,
+	      int *);
+UK_TRACEPOINT(trace_posix_socket_socketpair_ret, "%d", int);
+UK_TRACEPOINT(trace_posix_socket_socketpair_err, "%d", int);
+
+UK_SYSCALL_R_DEFINE(int, socketpair, int, family, int, type, int, protocol,
+	int *, usockfd)
+{
+	struct posix_socket_driver *d;
+	void *usockdata[2];
+	int vfs_fd1, vfs_fd2;
+	int ret;
+
+	trace_posix_socket_socketpair(family, type, protocol, usockfd);
+
+	d = posix_socket_driver_get(family);
+	if (unlikely(d == NULL)) {
+		ret = -EAFNOSUPPORT;
+		goto EXIT_ERR;
+	}
+
+	/* Create the socketpair using the driver */
+	ret = posix_socket_socketpair(d, family, type, protocol, usockdata);
+	if (unlikely(ret < 0))
+		goto EXIT_ERR;
+
+	/* Allocate the file descriptors */
+	vfs_fd1 = posix_socket_alloc_fd(d, type, usockdata[0]);
+	if (unlikely(vfs_fd1 < 0)) {
+		ret = PTR2ERR(vfs_fd1);
+		goto EXIT_ERR_CLOSE;
+	}
+
+	vfs_fd2 = posix_socket_alloc_fd(d, type, usockdata[1]);
+	if (unlikely(vfs_fd2 < 0)) {
+		ret = PTR2ERR(vfs_fd2);
+		goto EXIT_ERR_FREE_FD1;
+	}
+
+	usockfd[0] = vfs_fd1;
+	usockfd[1] = vfs_fd2;
+
+	trace_posix_socket_socketpair_ret(ret);
+	return ret;
+EXIT_ERR_FREE_FD1:
+	/* TODO: Broken error handling. Leaking vfs_fd1! */
```

Comment:
Why can't `vfs_fd1` be freed? Isn't there a `free()` equivalent to the `posix_socket_alloc_fd()` function?